src/frontend/src/binder/mod.rs

```diff
@@ -90,10 +93,78 @@ pub struct Binder {
 
     /// `ShareId`s identifying shared views.
     shared_views: HashMap<ViewId, ShareId>,
+
+    param_types: ParameterTypes,
+}
+
+#[derive(Clone, Debug)]
+pub struct ParameterTypes(Arc<RwLock<HashMap<u64, Option<DataType>>>>);
```

Comment:
As mentioned before, let's add some doc comments on:
* Why it exists (to collect inferred types when first `new_cast`, which does not have extra input / output)
* What does it mean when (a) not in the HashMap, (b) contains a `None`, (c) contains `Some(T)`